# ncbi-var-services

Variation Services
- API version: 2.0.0
  - Build date: 2019-04-25T10:32:24.325-07:00[America/Los_Angeles]

Services for variation data processing<p> <div style=\"font-size:1.1em;font-weight:normal\">NOTE: This service is still in beta testing mode. Please limit your request rate to 1 request/second.  dbSNP provides bulk download in VCF and JSON on the <a href=\"ftp://ftp.ncbi.nih.gov/snp/latest_release/\">FTP site</a> for users with a large number of RefSNPs (>100K) to process.</div> 


*Automatically generated by the [Swagger Codegen](https://github.com/swagger-api/swagger-codegen)*


## Requirements

Building the API client library requires:
1. Java 1.7+
2. Maven/Gradle

## Installation

To install the API client library to your local Maven repository, simply execute:

```shell
mvn clean install
```

To deploy it to a remote Maven repository instead, configure the settings of the repository and execute:

```shell
mvn clean deploy
```

Refer to the [OSSRH Guide](http://central.sonatype.org/pages/ossrh-guide.html) for more information.

### Maven users

Add this dependency to your project's POM:

```xml
<dependency>
  <groupId>org.pharmgkb</groupId>
  <artifactId>ncbi-var-services</artifactId>
  <version>1.0.0</version>
  <scope>compile</scope>
</dependency>
```

### Gradle users

Add this dependency to your project's build file:

```groovy
compile "org.pharmgkb:ncbi-var-services:1.0.0"
```

### Others

At first generate the JAR by executing:

```shell
mvn clean package
```

Then manually install the following JARs:

* `target/ncbi-var-services-1.0.0.jar`
* `target/lib/*.jar`

## Getting Started

Please follow the [installation](#installation) instruction and execute the following Java code:

```java
import org.pharmgkb.io.ncbi.*;
import org.pharmgkb.io.ncbi.auth.*;
import org.pharmgkb.io.ncbi.model.*;
import org.pharmgkb.io.ncbi.api.FrequencyNotYetAvailableComingSoonApi;

import java.io.File;
import java.util.*;

public class FrequencyNotYetAvailableComingSoonApiExample {

    public static void main(String[] args) {
        
        FrequencyNotYetAvailableComingSoonApi apiInstance = new FrequencyNotYetAvailableComingSoonApi();
        String seqId = "seqId_example"; // String | Sequence id in 'accession.version' format
        Integer position = 56; // Integer | Zero-based SPDI-like start position. The number of nucleotides to advance from the position before the first nucleotides before starting the interval.
        Integer length = 56; // Integer | SPDI-like length of interval in units of nucleotides
        try {
            Object result = apiInstance.frequencyByInterval(seqId, position, length);
            System.out.println(result);
        } catch (ApiException e) {
            System.err.println("Exception when calling FrequencyNotYetAvailableComingSoonApi#frequencyByInterval");
            e.printStackTrace();
        }
    }
}
import org.pharmgkb.io.ncbi.*;
import org.pharmgkb.io.ncbi.auth.*;
import org.pharmgkb.io.ncbi.model.*;
import org.pharmgkb.io.ncbi.api.FrequencyNotYetAvailableComingSoonApi;

import java.io.File;
import java.util.*;

public class FrequencyNotYetAvailableComingSoonApiExample {

    public static void main(String[] args) {
        
        FrequencyNotYetAvailableComingSoonApi apiInstance = new FrequencyNotYetAvailableComingSoonApi();
        try {
            Map result = apiInstance.frequencyMetadata();
            System.out.println(result);
        } catch (ApiException e) {
            System.err.println("Exception when calling FrequencyNotYetAvailableComingSoonApi#frequencyMetadata");
            e.printStackTrace();
        }
    }
}
```

## Documentation for API Endpoints

All URIs are relative to *https://api.ncbi.nlm.nih.gov/variation/v0*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*FrequencyNotYetAvailableComingSoonApi* | [**frequencyByInterval**](docs/FrequencyNotYetAvailableComingSoonApi.md#frequencyByInterval) | **GET** /interval/{seq_id}:{position}:{length}/overlapping_frequency_records | Retrieve frequencies for variants overlapping a position range
*FrequencyNotYetAvailableComingSoonApi* | [**frequencyMetadata**](docs/FrequencyNotYetAvailableComingSoonApi.md#frequencyMetadata) | **GET** /metadata/frequency | Retrieve frequency metadata
*HgvsApi* | [**hgvsBatchContextualsPost**](docs/HgvsApi.md#hgvsBatchContextualsPost) | **POST** /hgvs/batch/contextuals | Retrieve contextual alleles for each input
*HgvsApi* | [**hgvsHgvsContextualsGet**](docs/HgvsApi.md#hgvsHgvsContextualsGet) | **GET** /hgvs/{hgvs}/contextuals | Retrieve the contextual alleles equivalent to the HGVS notation input
*RefSnpApi* | [**betaRefsnpRsidGet**](docs/RefSnpApi.md#betaRefsnpRsidGet) | **GET** /beta/refsnp/{rsid} | Lookup RefSNP object by RSID
*SpdiApi* | [**spdiSpdiAllEquivalentContextualGet**](docs/SpdiApi.md#spdiSpdiAllEquivalentContextualGet) | **GET** /spdi/{spdi}/all_equivalent_contextual | Retrieve equivalent contextual allele on all aligned sequences
*SpdiApi* | [**spdiSpdiCanonicalRepresentativeGet**](docs/SpdiApi.md#spdiSpdiCanonicalRepresentativeGet) | **GET** /spdi/{spdi}/canonical_representative | Retrieve canonical allele representative
*SpdiApi* | [**spdiSpdiContextualGet**](docs/SpdiApi.md#spdiSpdiContextualGet) | **GET** /spdi/{spdi}/contextual | Retrieve contextual allele in SPDI syntax for provided allele
*SpdiApi* | [**spdiSpdiHgvsGet**](docs/SpdiApi.md#spdiSpdiHgvsGet) | **GET** /spdi/{spdi}/hgvs | Retrieve HGVS notation for contextual allele in SPDI Syntax
*SpdiApi* | [**spdiSpdiRsidsGet**](docs/SpdiApi.md#spdiSpdiRsidsGet) | **GET** /spdi/{spdi}/rsids | Lookup the RSIDs (if any) associated with the input allele
*SpdiApi* | [**spdiSpdiVcfFieldsGet**](docs/SpdiApi.md#spdiSpdiVcfFieldsGet) | **GET** /spdi/{spdi}/vcf_fields | Retrieve fields suitable for representing the input allele in VCF
*VcfApi* | [**vcfChromPosRefAltsContextualsGet**](docs/VcfApi.md#vcfChromPosRefAltsContextualsGet) | **GET** /vcf/{chrom}/{pos}/{ref}/{alts}/contextuals | Retrieve contextual alleles for the reference allele and each alternate allele
*VcfApi* | [**vcfFileSetRsidsPost**](docs/VcfApi.md#vcfFileSetRsidsPost) | **POST** /vcf/file/set_rsids | For provided VCF data, add RSIDs in the ID field

## Documentation for Models

 - [Allele](docs/Allele.md)
 - [AlleleAnnotationList](docs/AlleleAnnotationList.md)
 - [AssemblyAnnotForAllele](docs/AssemblyAnnotForAllele.md)
 - [ClinvarData](docs/ClinvarData.md)
 - [CodonAlignedTranscriptChange](docs/CodonAlignedTranscriptChange.md)
 - [Dbsnp1MergeEvent](docs/Dbsnp1MergeEvent.md)
 - [DetailedError](docs/DetailedError.md)
 - [Error](docs/Error.md)
 - [ErrorError](docs/ErrorError.md)
 - [ExternalRef](docs/ExternalRef.md)
 - [Frameshift](docs/Frameshift.md)
 - [FrequencyData](docs/FrequencyData.md)
 - [GeneAnnotForAllele](docs/GeneAnnotForAllele.md)
 - [Hgvs](docs/Hgvs.md)
 - [HgvsBatch](docs/HgvsBatch.md)
 - [HgvsBatchItem](docs/HgvsBatchItem.md)
 - [HgvsBatchSuccess](docs/HgvsBatchSuccess.md)
 - [HgvsSuccess](docs/HgvsSuccess.md)
 - [LostObsMovement](docs/LostObsMovement.md)
 - [MergedSnapshotData](docs/MergedSnapshotData.md)
 - [NosnpposSnapshotData](docs/NosnpposSnapshotData.md)
 - [PlacementAnnot](docs/PlacementAnnot.md)
 - [PlacementAnnotatedAllele](docs/PlacementAnnotatedAllele.md)
 - [PlacementWithAllele](docs/PlacementWithAllele.md)
 - [PresentObsMovement](docs/PresentObsMovement.md)
 - [PreviousRelease](docs/PreviousRelease.md)
 - [PrimarySnapshotData](docs/PrimarySnapshotData.md)
 - [ProteinAnnotForAllele](docs/ProteinAnnotForAllele.md)
 - [ProteinVariant](docs/ProteinVariant.md)
 - [RefsnpSnapshot](docs/RefsnpSnapshot.md)
 - [RefsnpSnapshotSuccess](docs/RefsnpSnapshotSuccess.md)
 - [RnaAnnotForAllele](docs/RnaAnnotForAllele.md)
 - [Rsids](docs/Rsids.md)
 - [RsidsSuccess](docs/RsidsSuccess.md)
 - [SeqIdTraitByAssembly](docs/SeqIdTraitByAssembly.md)
 - [SoTerm](docs/SoTerm.md)
 - [Spdi](docs/Spdi.md)
 - [SpdiSuccess](docs/SpdiSuccess.md)
 - [Spdis](docs/Spdis.md)
 - [SpdisSuccess](docs/SpdisSuccess.md)
 - [Support](docs/Support.md)
 - [SupportId](docs/SupportId.md)
 - [UnsupportedSnapshotData](docs/UnsupportedSnapshotData.md)
 - [VcfFields](docs/VcfFields.md)
 - [VcfFieldsSuccess](docs/VcfFieldsSuccess.md)
 - [Warnings](docs/Warnings.md)
 - [WithdrawnSnapshotData](docs/WithdrawnSnapshotData.md)

## Documentation for Authorization

All endpoints do not require authorization.
Authentication schemes defined for the API:

## Recommendation

It's recommended to create an instance of `ApiClient` per thread in a multithreaded environment to avoid any potential issues.

## Author

variation-services@ncbi.nlm.nih.gov
