/*
 * Variation Services
 * Services for variation data processing<p> <div style=\"font-size:1.1em;font-weight:normal\">NOTE: This service is still in beta testing mode. Please limit your request rate to 1 request/second.  dbSNP provides bulk download in VCF and JSON on the <a href=\"ftp://ftp.ncbi.nih.gov/snp/latest_release/\">FTP site</a> for users with a large number of RefSNPs (>100K) to process.</div> 
 *
 * OpenAPI spec version: 2.0.0
 * Contact: variation-services@ncbi.nlm.nih.gov
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

package org.pharmgkb.io.ncbi.model;

import java.util.Objects;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import io.swagger.v3.oas.annotations.media.Schema;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.pharmgkb.io.ncbi.model.Spdi;

/**
 * Data specific to unsupported snapshots; cannot be present at the same time as the other *_snapshot_data fields
 */
@Schema(description = "Data specific to unsupported snapshots; cannot be present at the same time as the other *_snapshot_data fields")
@javax.annotation.Generated(value = "io.swagger.codegen.v3.generators.java.JavaClientCodegen", date = "2019-04-25T10:32:24.325-07:00[America/Los_Angeles]")public class UnsupportedSnapshotData {

  @SerializedName("alleles_from_previous_support")
  private List<Spdi> allelesFromPreviousSupport = new ArrayList<Spdi>();
  public UnsupportedSnapshotData allelesFromPreviousSupport(List<Spdi> allelesFromPreviousSupport) {
    this.allelesFromPreviousSupport = allelesFromPreviousSupport;
    return this;
  }

  public UnsupportedSnapshotData addAllelesFromPreviousSupportItem(Spdi allelesFromPreviousSupportItem) {
    this.allelesFromPreviousSupport.add(allelesFromPreviousSupportItem);
    return this;
  }

  /**
  * The set of alleles that this RefSnp had in the build just prior to becoming unsupported.
  * @return allelesFromPreviousSupport
  **/
  @Schema(required = true, description = "The set of alleles that this RefSnp had in the build just prior to becoming unsupported.")
  public List<Spdi> getAllelesFromPreviousSupport() {
    return allelesFromPreviousSupport;
  }
  public void setAllelesFromPreviousSupport(List<Spdi> allelesFromPreviousSupport) {
    this.allelesFromPreviousSupport = allelesFromPreviousSupport;
  }
  @Override
  public boolean equals(java.lang.Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    UnsupportedSnapshotData unsupportedSnapshotData = (UnsupportedSnapshotData) o;
    return Objects.equals(this.allelesFromPreviousSupport, unsupportedSnapshotData.allelesFromPreviousSupport);
  }

  @Override
  public int hashCode() {
    return java.util.Objects.hash(allelesFromPreviousSupport);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class UnsupportedSnapshotData {\n");
    
    sb.append("    allelesFromPreviousSupport: ").append(toIndentedString(allelesFromPreviousSupport)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(java.lang.Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }

}
